src/OpenLoco/src/GameCommands/CreateTown.cpp

```diff
@@ -0,0 +1,200 @@
+        regs.ax = pos.x;
+        regs.cx = pos.y;
+        call(0x004C5604, regs);
+        return regs.dx;
+    }
+
+    static bool checkSurroundings(Pos2 pos)
+    {
+        auto tile = TileManager::get(pos);
+        auto* surfaceEl = tile.surface();
+        if (surfaceEl->slope() != 0 || surfaceEl->water() != 0)
+        {
+            return false;
+        }
+
+        auto* landObj = ObjectManager::get<LandObject>(surfaceEl->terrain());
+        if (landObj->hasFlags(LandObjectFlags::isDesert | LandObjectFlags::noTrees))
+        {
+            auto dx = sub_4C5604(pos);
+            if (dx < 10 && getGameState().rng.randNext() & 0xFF)
+            {
+                return false;
+            }
+        }
+
+        for (auto& town : TownManager::towns())
+        {
+            auto manhattanDistance = Math::Vector::manhattanDistance(pos, Pos2(town.x, town.y));
+            if (manhattanDistance < 768)
+            {
+                setErrorText(StringIds::too_close_to_another_town);
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    // 0x00496C22
+    static uint32_t createTown(const TownPlacementArgs& args, const uint8_t flags)
+    {
+        auto& gameState = getGameState();
+        Pos2 pos = args.pos;
+        Town* newTown = nullptr;
+
+        if (args.pos.x == -1)
+        {
+            bool foundPos = false;
+            for (auto attempts = 200; attempts > 0; attempts--)
+            {
+                uint32_t rand = gameState.rng.randNext();
+                auto tilePos = TilePos2(((rand >> 16) * kMapColumns) >> 16, ((rand & 0xFFFF) * kMapRows) >> 16);
+                pos = toWorldSpace(tilePos);
+
+                if (pos.x < 384 || pos.y < 384 || pos.x > 11904 || pos.y > 11904)
+                {
+                    continue;
+                }
+
+                if (checkSurroundings(pos))
+                {
+                    foundPos = true;
+                    break;
+                }
+            }
+
+            if (!foundPos)
+            {
+                return FAILURE;
+            }
+        }
+        else
+        {
+            bool foundPos = false;
+            for (int attempts = 0; attempts < 40; attempts++)
+            {
+                if (attempts == 0)
+                {
+                    uint32_t rand = gameState.rng.randNext();
+                    pos += Pos2(((rand >> 4) & 0xE0) - 0x60, (rand & 0xE0) - 0x60);
+                }
+                else if (attempts <= 10)
+                {
+                    uint32_t rand = gameState.rng.randNext();
+                    pos += Pos2(((rand >> 4) & 0x60) - kTileSize, (rand & 0x60) - kTileSize);
+                }
+
+                if (pos.x < 160 || pos.y < 160 || pos.x > 12128 || pos.y > 12128)
+                {
+                    continue;
+                }
+
+                if (checkSurroundings(pos))
+                {
+                    foundPos = true;
```

Comment:
If the above changes are made you'll need to copy the attemptLoc into pos at this point.